src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,100 @@ public ProjectGraph(string entryProjectFile)
-                    foreach (var projectReferenceToParse in projectReferenceItems)
+                    // a project reference can be in "Processed" state. If it is "InProcess" state, it is an ancestor and there is a circular dependency
+                    if (projectReferenceNodeState == NodeState.InProcess)
                     {
-                        if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                        if (projectToEvaluate.ProjectFullPath.Equals(projectReferenceConfigurationMetadata.ProjectFullPath))
                         {
-                            throw new InvalidOperationException(string.Format(
-                                CultureInfo.InvariantCulture,
-                                ResourceUtilities.GetResourceString("ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
-                                projectReferenceToParse.EvaluatedInclude,
-                                parsedProject.Project.FullPath));
+                            // the project being evaluated has a reference on itself
+                            throw new CircularDependencyException(string.Format(
+                                ResourceUtilities.GetResourceString("CircularDependencyInProjectGraph"),
+                                projectToEvaluate.ProjectFullPath));
+                        }
+                        else
+                        {
+                            // the project being evaluated has a circular dependency involving multiple projects
+                            // add this project to the list of projects involved in cycle 
+                            var projectsInCycle = new List<string> {projectReferenceConfigurationMetadata.ProjectFullPath};
+                            return (false, projectsInCycle);
+                        }
+                    }
+                }
+                else
+                {
+                    // a new project that has to be evaluated
+                    var loadReference = LoadGraph(projectReferenceConfigurationMetadata, nodeState, projectCollection,
+                        globalProperties);
+                    if (!loadReference.success)
+                    {
+                        if (loadReference.projectsInCycle[0].Equals(parsedProject.Project.FullPath))
+                        {
+                            // we have reached the nth project in the cycle, form error message and throw
+                            loadReference.projectsInCycle.Add(projectReferenceConfigurationMetadata.ProjectFullPath);
+                            var errorMessage = new StringBuilder(500);
+                            for (int i = loadReference.projectsInCycle.Count-1; i >= 0; i--)
+                            {
+                                errorMessage.Append(loadReference.projectsInCycle[i]).AppendLine();
```

Comment:
Instead of newlines, would something like "->" make more sense (although you'd need to account for the leading/trailing arrow.

The error showing: `2.proj -> 6.proj -> 7.proj -> 3.proj -> 2.proj` seems super clear as to what the cycle is and in what direction.

Reply:
Or, if with full file paths the string gets really long, maybe prepend the lines with an error?

I'm not really feeling super strongly about this suggestion though. Maybe look at a real error message and use a judgement call.

Reply:
This is how the final error message looks like:

MSB4251: There is a circular dependency involving the following projects: 
C:\Users\yaupadh\AppData\Local\Temp\csfr4ykv.2te\Temporaryb33fd5235aa84433ad2e1717bf71add3\2.proj ->
C:\Users\yaupadh\AppData\Local\Temp\csfr4ykv.2te\Temporaryb33fd5235aa84433ad2e1717bf71add3\6.proj ->
C:\Users\yaupadh\AppData\Local\Temp\csfr4ykv.2te\Temporaryb33fd5235aa84433ad2e1717bf71add3\7.proj ->
C:\Users\yaupadh\AppData\Local\Temp\csfr4ykv.2te\Temporaryb33fd5235aa84433ad2e1717bf71add3\3.proj ->
C:\Users\yaupadh\AppData\Local\Temp\csfr4ykv.2te\Temporaryb33fd5235aa84433ad2e1717bf71add3\2.proj